airstack-modules/modules/airstack/domain-name/domain-name.ts

```diff
@@ -0,0 +1,1074 @@
+    if (domain.owner == getOrCreateAirAccount(chainId, ZERO_ADDRESS).id && domain.subdomainCount == BIG_INT_ZERO) {
+      if (domain.parent) {
+        const parentDomain = getOrCreateAirDomain(new Domain(domain.parent!, chainId, block, tokenAddress));
+        if (parentDomain) {
+          parentDomain.subdomainCount = parentDomain.subdomainCount.minus(BIGINT_ONE)
+          parentDomain.lastBlock = block.id;
+          parentDomain.save();
+          return recurseDomainDelete(parentDomain, chainId, block, tokenAddress)
+        }
+      }
+      return null
+    }
+    return domain.id
+  }
+
+  /**
+   * @dev this function updates air entity counter for a given entity id
+   * @param id entity id for entity to be updated
+   * @param block air block object
+   * @returns updated entity count
+   */
+  function updateAirEntityCounter(
+    id: string,
+    block: AirBlock,
+  ): BigInt {
+    let entity = AirEntityCounter.load(id);
+    if (entity == null) {
+      entity = new AirEntityCounter(id);
+      entity.count = BIGINT_ONE;
+      entity.createdAt = block.id;
+      entity.lastUpdatedAt = block.id;
+      createAirMeta(SUBGRAPH_SLUG, SUBGRAPH_NAME);
+    } else {
+      entity.count = entity.count.plus(BIGINT_ONE);
+      entity.lastUpdatedAt = block.id;
+    }
+    entity.save();
+    return entity.count as BigInt;
+  }
+
+  /**
+   * @dev this function creates air meta entity
+   * @param slug subgraph slug
+   * @param name subgraph name
+   */
+  function createAirMeta(
+    slug: string,
+    name: string
+    // should ideally have version also being passed from here
+  ): void {
+    let meta = AirMeta.load(AIR_META_ID);
+    if (meta == null) {
+      meta = new AirMeta(AIR_META_ID);
+      meta.network = processNetwork(dataSource.network());
+      meta.schemaVersion = SUBGRAPH_SCHEMA_VERSION;
+      meta.version = SUBGRAPH_VERSION;
+      meta.slug = slug;
+      meta.name = name;
+      meta.save();
+    }
+  }
+
+  /**
+   * @dev this function gets or creates a new air block entity
+   * @param chainId chain id
+   * @param blockHeight block number
+   * @param blockHash block hash
+   * @param blockTimestamp block timestamp
+   * @returns AirBlock entity
+   */
+  function getOrCreateAirBlock(
+    chainId: string,
+    blockHeight: BigInt,
+    blockHash: string,
+    blockTimestamp: BigInt
+  ): AirBlock {
+    let id = chainId.concat("-").concat(blockHeight.toString());
+
+    let block = AirBlock.load(id);
+    if (block == null) {
+      block = new AirBlock(id);
+      block.hash = blockHash;
+      block.number = blockHeight;
+      block.timestamp = blockTimestamp
+      block.save()
+    }
+    return block as AirBlock;
+  }
+
+  /**
+   * @dev this function gets or creates a new air token entity
+   * @param chainID chain id
+   * @param address token address
+   * @returns AirToken entity
+   */
+  function getOrCreateAirToken(chainID: string, address: string): AirToken {
```

Comment:
Move to a common file.

Reply:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate air token

Reply:
also, nfs-marketplace and domain-name verticals have a different schema for air token

Reply:
cannot do, coz air token in domain-name vertical only has id and address fields, while nft-marketplace vertical has a lot of other fields also